pkg/tsdb/azuremonitor/metrics/url-builder.go

```diff
@@ -7,33 +7,50 @@ import (
 
 // urlBuilder builds the URL for calling the Azure Monitor API
 type urlBuilder struct {
+	ResourceURI string
+
+	// Following fields will be deprecated in grafana 9 and will not included in new queries.
+	// For backwards compat, we recreate the ResourceURI using these fields
 	DefaultSubscription string
 	Subscription        string
 	ResourceGroup       string
 	MetricDefinition    string
 	ResourceName        string
 }
 
-// Build checks the metric definition property to see which form of the url
+// BuildMetricsURL checks the metric definition property to see which form of the url
 // should be returned
-func (ub *urlBuilder) Build() string {
-	subscription := ub.Subscription
-
-	if ub.Subscription == "" {
-		subscription = ub.DefaultSubscription
-	}
-
-	metricDefinitionArray := strings.Split(ub.MetricDefinition, "/")
-	resourceNameArray := strings.Split(ub.ResourceName, "/")
-	provider := metricDefinitionArray[0]
-	metricDefinitionArray = metricDefinitionArray[1:]
-
-	urlArray := []string{subscription, "resourceGroups", ub.ResourceGroup, "providers", provider}
-
-	for i := range metricDefinitionArray {
-		urlArray = append(urlArray, metricDefinitionArray[i])
-		urlArray = append(urlArray, resourceNameArray[i])
+func (params *urlBuilder) BuildMetricsURL() string {
+	resourceURI := params.ResourceURI
+
+	// Prior to Grafana 9, we had a legacy query object rather than a resourceURI, so we manually create the resource URI
+	if resourceURI == "" {
```

Comment:
Oh and all of this block could be extracted to a function with the name "manuallyCreateResourceURI" for example.

(I personally like when the code explains itself without comments. Comments are often overlooked/ignored/they even are super light grey here and in my IDE... and they're often not updated whereas the code will be.)